next/frontend/utils/queryParamRedirect.ts

```diff
@@ -0,0 +1,151 @@
+}
+
+export const redirectQueryParam = 'from'
+
+export type SafeRedirect = {
+  url: string
+  type: SafeRedirectType
+}
+
+const defaultSafeRedirect: SafeRedirect = {
+  url: ROUTES.HOME,
+  type: SafeRedirectType.Local,
+}
+
+export const isDefaultRedirect = (loginRedirectResult: SafeRedirect) =>
+  loginRedirectResult.url === ROUTES.HOME
+
+/**
+ * Returns a safe variant (including type) of the redirect query param. For remote URLs, we only allow a subset of origins.
+ */
+export const getSafeRedirect = (queryParam: unknown) => {
+  if (typeof queryParam !== 'string') {
+    return defaultSafeRedirect
+  }
+
+  if (queryParam.startsWith('/')) {
+    return {
+      url: queryParam,
+      type: SafeRedirectType.Local,
+    }
+  }
+
+  try {
+    const url = new URL(queryParam)
+    if (APPROVED_LOGIN_REDIRECT_ORIGINS.includes(url.origin)) {
+      return {
+        url: queryParam,
+        type: SafeRedirectType.Remote,
+      }
+    }
+    // eslint-disable-next-line no-empty
+  } catch (error) {}
+
+  return defaultSafeRedirect
+}
+
+/**
+ * We want to keep only viable and safe query params in the URL, otherwise it is removed.
+ */
+export const shouldRemoveRedirectQueryParam = (originalQueryParam: unknown) => {
+  if (typeof originalQueryParam !== 'string') {
+    return false
+  }
+
+  const safeRedirect = getSafeRedirect(originalQueryParam)
+  if (isDefaultRedirect(safeRedirect)) {
+    return true
+  }
+
+  // If the "safe redirect" is different from the original query param, we want to remove the original query param
+  return safeRedirect?.url !== originalQueryParam
+}
+
+/**
+ * Parses the resolved URL and removes the redirect query param if it exists.
+ *
+ * @example
+ * /some-path?from=/some-other-path -> /some-path
+ * /some-path?from=/some-other-path&other=param -> /some-path?other=param
+ * /some-path -> /some-path
+ */
+export const removeRedirectQueryParamFromUrl = (resolvedUrl: string) => {
+  const splitUrl = resolvedUrl.split('?')
+  if (splitUrl.length === 1) {
+    return resolvedUrl
+  }
+
+  const [url, ...queryStringArray] = splitUrl
+  const queryString = queryStringArray.join('')
+
+  const queryParams = qs.parse(queryString)
+  delete queryParams[redirectQueryParam]
+  const newQueryString = qs.stringify(queryParams)
+
+  if (newQueryString) {
+    return `${url}?${newQueryString}`
+  }
+  return url
+}
+
+/**
+ * Returns the URL to redirect to. Remote URLs in redirect query params are used to pass the access token to the target
+ * domain, therefore we add the access token to the query string.
+ */
+export const getRedirectUrl = async (
+  safeRedirect: SafeRedirect,
+  getAccessToken: () => Promise<string | null>,
+) => {
+  if (safeRedirect.type === SafeRedirectType.Remote) {
+    const parsedUrl = new URL(safeRedirect.url)
+    // add access token to the query string
+    const accessToken = await getAccessToken()
+    if (accessToken) {
+      parsedUrl.searchParams.set('access_token', accessToken)
+    }
```

Comment:
would be super happy to get rid of this if we have the opportunity to do so now. It would require something along the lines of the cookie authentication on bratislava.sk domain that is open in the other PR.

Reply:
Yes, please elaborate in Slack or somewhere, I don't know yet how this should be handled properly, so I just copied the functionality.